homeassistant/components/reolink/binary_sensor.py

```diff
@@ -0,0 +1,172 @@
+"""This component provides support for Reolink binary sensors."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from reolink_aio.api import (
+    FACE_DETECTION_TYPE,
+    PERSON_DETECTION_TYPE,
+    PET_DETECTION_TYPE,
+    VEHICLE_DETECTION_TYPE,
+)
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ReolinkData
+from .const import DOMAIN
+from .entity import ReolinkCoordinatorEntity
+from .host import ReolinkHost
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class ReolinkBinarySensorDescription(BinarySensorEntityDescription):
+    """A class that describes binary sensor entities."""
+
+    icon: str = "mdi:motion-sensor"
+    icon_off: str = "mdi:motion-sensor-off"
+    value: Callable[[ReolinkHost, int | None], bool | None] = lambda host, ch: None
+    supported: Callable[[ReolinkHost, int | None], bool] = lambda host, ch: True
+
+
+BINARY_SENSORS = (
+    ReolinkBinarySensorDescription(
+        key="motion",
+        name="Motion",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda host, ch: host.api.motion_detected(ch),
+    ),
+    ReolinkBinarySensorDescription(
+        key=FACE_DETECTION_TYPE,
+        name=FACE_DETECTION_TYPE,
+        icon="mdi:face-recognition",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda host, ch: host.api.ai_detected(ch, FACE_DETECTION_TYPE),
```

Comment:
The method `ai_detected` can return either a `bool` or a `dict[str, bool]`, see:

![image](https://user-images.githubusercontent.com/195327/212780050-1367e7bb-8f4c-481c-a92b-6bd4320abe0d.png)

The latter `dict[str, bool]` hasn't been taken into account in type or handling.

Reply:
```suggestion
        value=lambda api, ch: api.ai_detected(ch, FACE_DETECTION_TYPE),
```

Reply:
We are hard coding the object_type as a input which is not None, therefore the last return you mention can never be reached, so it is not a problem.
I do agree that typing in the upstream lib is a bit sloppy, I will look into this.

Reply:
I'm unresolving this, as it is a problem. Typing isn't correct.

Typing says it can be returned, so it should be handled, or typing should be fixed.

Reply:
![CleanShot 2023-01-17 at 01 11 53](https://user-images.githubusercontent.com/195327/212783169-da0c66a9-9e6e-4ee4-8ab9-0f14c97d79bb.png)

Reply:
Alright, I updated the upstream library main branch to fix the return type to a bool and split out the diffrent method ai_detection_states to get the dictionary:
https://github.com/starkillerOG/reolink_aio/commit/f6fcdf0a22c060fff447e73be10766d8d6c9c310

Do I really have to do a version release and bump the HomeAssistant reolink-aio version just for this typing?

Reply:
Version 0.3.0 of reolink-aio has been released and the version bump PR is here: https://github.com/home-assistant/core/pull/86259.
That will fix this typing, therefore resolving this issue.